aws/logs_monitoring/cache.py

```diff
@@ -435,10 +440,141 @@ def get(self, log_group):
                 logger.debug(
-                    "Not fetching custom tags because the env variable DD_FETCH_LAMBDA_TAGS is not set to true"
+                    "Not fetching custom tags because the env variable DD_FETCH_LOG_GROUP_TAGS is "
+                    "not set to true"
                 )
                 return []
             log_group_tags = get_log_group_tags(log_group) or []
             self.tags_by_id[log_group] = log_group_tags
 
         return log_group_tags
+
+
+#######################
+# Step Functions Tags #
+#######################
+
+
+class StepFunctionsTagsCache(LambdaTagsCache):
+    CACHE_FILENAME = DD_S3_STEP_FUNCTIONS_CACHE_FILENAME
+    CACHE_LOCK_FILENAME = DD_S3_STEP_FUNCTIONS_CACHE_LOCK_FILENAME
+
+    def should_fetch_tags(self):
+        return should_fetch_step_functions_tags()
+
+    def build_tags_cache(self):
+        """Makes API calls to GetResources to get the live tags of the account's Step Functions
+        Returns an empty dict instead of fetching custom tags if the tag fetch env variable is not
+        set to true.
+        Returns:
+            tags_by_arn_cache (dict<str, str[]>): each Lambda's tags in a dict keyed by ARN
+        """
+        tags_fetch_success = False
+        tags_by_arn_cache = {}
+        get_resources_paginator = resource_tagging_client.get_paginator("get_resources")
+
+        try:
+            for page in get_resources_paginator.paginate(
+                ResourceTypeFilters=["states"], ResourcesPerPage=100
+            ):
+                send_forwarder_internal_metrics(
+                    "step_functions_get_resources_api_calls"
+                )
+                page_tags_by_arn = parse_get_resources_response_for_tags_by_arn(page)
+                tags_by_arn_cache.update(page_tags_by_arn)
+                tags_fetch_success = True
+
+        except ClientError as e:
+            logger.exception(
+                "Encountered a ClientError when trying to fetch tags. You may need to give "
+                "this Lambda's role the 'tag:GetResources' permission"
+            )
+            additional_tags = [
+                f"http_status_code:{e.response['ResponseMetadata']['HTTPStatusCode']}"
+            ]
+            send_forwarder_internal_metrics(
+                "client_error", additional_tags=additional_tags
+            )
+
+        logger.debug("All Step Functions tags refreshed: {}".format(tags_by_arn_cache))
+
+        return tags_fetch_success, tags_by_arn_cache
+
+    def get(self, state_machine_arn):
+        """Get the tags for the Step Functions from the cache
+
+        Will re-fetch the tags if they are out of date, or a log group is encountered
+        which isn't in the tag list
+
+        Args:
+            state_machine_arn (str): the key we're getting tags from the cache for
+
+        Returns:
+            state_machine_tags (List[str]): the list of "key:value" Datadog tag strings
+        """
+        if self._is_expired():
```

Comment:
Just curious. So, for a specific execution_arn, its tags are keeping changing? What kind of change can it have? Or once its tags are not null, there will be no change for these tags.

Reply:
tags are cached for 300 seconds. Every 300 seconds, forwarder will refetch all tag for state machines. Does this answer your question?